sled-agent/config-reconciler/src/reconciler_task.rs

```diff
@@ -127,21 +147,292 @@ pub enum ReconcilerTaskStatus {
+                            // in tests.
+                            warn!(
+                                self.log,
+                                "raw_disks watch channel closed; exiting"
+                            );
+                            return;
+                        }
+                    }
+                }
+
+                // Cancel-safe: this is either `future::pending()` (never
+                // completes) or `sleep()` (we don't care if it's cancelled)
+                _ = maybe_retry => {
+                    info!(
+                        self.log,
+                        "starting reconciliation due to retryable error"
+                    );
+                    continue;
+                }
+            }
+        }
     }
+
+    async fn do_reconcilation<T: SledAgentFacilities>(
+        &mut self,
+        sled_agent_facilities: &T,
+    ) -> ReconciliationResult {
+        // Take a snapshot of the current state of the input channels on which
+        // we act. Clone both to avoid keeping the channels locked while we
+        // reconcile.
+        let current_config = self.current_config_rx.borrow_and_update().clone();
+        let current_raw_disks = self.raw_disks_rx.borrow_and_update().clone();
+
+        // See whether we actually have a config to reconcile against.
+        let started_at_instant = Instant::now();
+        let sled_config = match current_config {
+            // In both `WaitingFor...` cases, we don't need to retry on our own:
+            // we'll retry as soon as there's a change in our inputs that might
+            // nudge us out of these states.
+            CurrentSledConfig::WaitingForInternalDisks => {
+                self.reconciler_result_tx.send_modify(|r| {
+                    r.status = ReconcilerTaskStatus::WaitingForInternalDisks;
+                });
+                return ReconciliationResult::NoRetryNeeded;
+            }
+            CurrentSledConfig::WaitingForInitialConfig => {
+                self.reconciler_result_tx.send_modify(|r| {
+                    r.status = ReconcilerTaskStatus::WaitingForInitialConfig;
+                });
+                return ReconciliationResult::NoRetryNeeded;
+            }
+            CurrentSledConfig::Ledgered(sled_config) => {
+                self.reconciler_result_tx.send_modify(|r| {
+                    r.status = ReconcilerTaskStatus::PerformingReconciliation {
+                        config: sled_config.clone(),
+                        started_at_time: Utc::now(),
+                        started_at_instant,
+                    };
+                });
+                sled_config
+            }
+        };
+
+        // ---
+        // We go through the removal process first: shut down zones, then remove
+        // datasets, then remove disks.
+        // ---
+
+        // First, shut down zones if needed.
+        let zone_shutdown_result = self
+            .zones
+            .shut_down_zones_if_needed(
```

Comment:
We're relying on the `PartiallyShutDown` ZoneState here, so that repeatedly calling this returns the same error until something changes, right?

Reply:
That's right. If a zone is in one of the `PartiallyShutDown` states when this returns, it will still be in that state when we get here again, and we'll try continuing from the step that failed.